File: java/structure/tree/SegmentTree.java
package structure.tree;

public class SegmentTree {
    private TreeNode[] _nodes = null;
    private int _length = 0;

    public SegmentTree(int[] nums) {
        _nodes = new TreeNode[nums.length << 2];
        for (int i = 0; i < (nums.length << 2); ++i) {
            _nodes[i] = new TreeNode();
        }
        _length = nums.length;
        createNode(nums, 0, nums.length - 1, 0);
    }

    public void set(int index, int value) {
        if (index >= 0 && index < _length) {
            setNode(index, value, 0);
        }
    }

    public void setInterval(int start, int end, int value) {
        if (start >= 0 && start < _length && end >= 0 && end < _length) {
            setNodes(start, end, value, 0);
        }
    }

    public int get(int index) {
        if (index >= 0 && index < _length) {
            return getNodes(index, index, 0);
        }
        return Integer.MIN_VALUE;
    }

    public int getInterval(int start, int end) {
        if (start >= 0 && start < _length && end >= 0 && end < _length) {
            return getNodes(start, end, 0);
        }
        return Integer.MIN_VALUE;
    }

    private void createNode(int[] nums, int start, int end, int parent) {
        _nodes[parent].start = start;
        _nodes[parent].end = end;
        if (start == end) {
            _nodes[parent].value = nums[start];
            return;
        }

        int middle = start + ((end - start) >> 1);
        int child = (parent << 1) + 1;
        createNode(nums, start, middle, child);
        createNode(nums, middle + 1, end, child + 1);
        _nodes[parent].value = _nodes[child].value + _nodes[child + 1].value;
    }

    private void setNode(int index, int value, int parent) {
        if (_nodes[parent].start == _nodes[parent].end) {
            _nodes[parent].value = value;
            return;
        }

        int middle = _nodes[parent].start + ((_nodes[parent].end - _nodes[parent].start) >> 1);
        int child = (parent << 1) + 1;
        if (index <= middle) {
            setNode(index, value, child);
        } else {
            setNode(index, value, child + 1);
        }
        _nodes[parent].value = _nodes[child].value + _nodes[child + 1].value;
    }

    private void setNodes(int start, int end, int value, int parent) {
        if (start == _nodes[parent].start && end == _nodes[parent].end) {
            _nodes[parent].value += (end - start + 1) * value;
            _nodes[parent].covered += value;
            return;
        }

        int middle = _nodes[parent].start + ((_nodes[parent].end - _nodes[parent].start) >> 1);
        int child = (parent << 1) + 1;
        if (end <= middle) {
            setNodes(start, end, value, child);
        } else if (start > middle) {
            setNodes(start, end, value, child + 1);
        } else {
            setNodes(start, middle, value, child);
            setNodes(middle + 1, end, value, child + 1);
        }
        _nodes[parent].value = _nodes[child].value + _nodes[child + 1].value;
    }

    private int getNodes(int start, int end, int parent) {
        if (0 != _nodes[parent].covered) {
            if (_nodes[parent].start != _nodes[parent].end) {
                int child = (parent << 1) + 1;
                int covered = _nodes[parent].covered;
                _nodes[child].covered += covered;
                _nodes[child + 1].covered += covered;
                if (0 != covered) {
                    _nodes[child].value += (_nodes[child].end - _nodes[child].start + 1) * covered;
                    _nodes[child + 1].value += (_nodes[child + 1].end - _nodes[child + 1].start + 1) * covered;
                }
            }
            _nodes[parent].covered = 0;
        }

        if (start == _nodes[parent].start && end == _nodes[parent].end) {
            return _nodes[parent].value;
        }

        int middle = _nodes[parent].start + ((_nodes[parent].end - _nodes[parent].start) >> 1);
        int child = (parent << 1) + 1;
        if (end <= middle) {
            return getNodes(start, end, child);
        }
        if (start > middle) {
            return getNodes(start, end, child + 1);
        }
        return getNodes(start, middle, child) + getNodes(middle + 1, end, child + 1);
    }

    private class TreeNode {
        public int start = 0;
        public int end = 0;
        public int value = 0;
        public int covered = 0;
    }
}

File: py/structure/tree/SegmentTree.py
#!/usr/bin/env python
# -*- coding:gbk -*-

import sys
from typing import List

class SegmentTree:
    def __init__(self, nums: List[int]):
        self.__nodes, self.__length = [self.__TreeNode() for _ in range((len(nums) << 2))], len(nums)
        self.__createNode(nums, 0, len(nums) - 1, 0)

    def set(self, index: int, value: int) -> None:
        if index >= 0 and index < self.__length:
            self.__setNode(index, value, 0)

    def setInterval(self, start: int, end: int, value: int) -> None:
        if start >= 0 and start < self.__length and end >= 0 and end < self.__length:
            self.__setNodes(start, end, value, 0)

    def get(self, index: int) -> int:
        if index >= 0 and index < self.__length:
            return self.__getNodes(index, index, 0)
        return -sys.maxsize

    def getInterval(self, start: int, end: int) -> int:
        if start >= 0 and start < self.__length and end >= 0 and end < self.__length:
            return self.__getNodes(start, end, 0)
        return -sys.maxsize

    def __createNode(self, nums: List[int], start: int, end: int, parent: int) -> None:
        self.__nodes[parent].start, self.__nodes[parent].end = start, end
        if start == end:
            self.__nodes[parent].value = nums[start]
            return

        middle, child = start + ((end - start) >> 1), (parent << 1) + 1
        self.__createNode(nums, start, middle, child)
        self.__createNode(nums, middle + 1, end, child + 1)
        self.__nodes[parent].value = self.__nodes[child].value + self.__nodes[child + 1].value

    def __setNode(self, index: int, value: int, parent: int) -> int:
        if self.__nodes[parent].start == self.__nodes[parent].end:
            self.__nodes[parent].value = value
            return

        middle, child = self.__nodes[parent].start + ((self.__nodes[parent].end - self.__nodes[parent].start) >> 1), (parent << 1) + 1
        if index <= middle:
            self.__setNode(index, value, child)
        else:
            self.__setNode(index, value, child + 1)
        self.__nodes[parent].value = self.__nodes[child].value + self.__nodes[child + 1].value

    def __setNodes(self, start: int, end: int, value: int, parent: int) -> int:
        if start == self.__nodes[parent].start and end == self.__nodes[parent].end:
            self.__nodes[parent].value += (end - start + 1) * value
            self.__nodes[parent].covered += value
            return

        middle, child = self.__nodes[parent].start + ((self.__nodes[parent].end - self.__nodes[parent].start) >> 1), (parent << 1) + 1
        if end <= middle:
            self.__setNodes(start, end, value, child)
        elif start > middle:
            self.__setNodes(start, end, value, child + 1)
        else:
            self.__setNodes(start, middle, value, child)
            self.__setNodes(middle + 1, end, value, child + 1)
        self.__nodes[parent].value = self.__nodes[child].value + self.__nodes[child + 1].value

    def __getNodes(self, start: int, end: int, parent: int) -> int:
        if self.__nodes[parent].covered:
            if self.__nodes[parent].start != self.__nodes[parent].end:
                child, covered = (parent << 1) + 1, self.__nodes[parent].covered
                self.__nodes[child].covered += covered
                self.__nodes[child + 1].covered += covered
                if covered:
                    self.__nodes[child].value += (self.__nodes[child].end - self.__nodes[child].start + 1) * covered
                    self.__nodes[child + 1].value += (self.__nodes[child + 1].end - self.__nodes[child + 1].start + 1) * covered
            self.__nodes[parent].covered = 0

        if start == self.__nodes[parent].start and end == self.__nodes[parent].end:
            return self.__nodes[parent].value

        middle, child = self.__nodes[parent].start + ((self.__nodes[parent].end - self.__nodes[parent].start) >> 1), (parent << 1) + 1
        if end <= middle:
            return self.__getNodes(start, end, child)
        if start > middle:
            return self.__getNodes(start, end, child + 1)
        return self.__getNodes(start, middle, child) + self.__getNodes(middle + 1, end, child + 1)

    class __TreeNode:
        def __init__(self):
            self.start, self.end, self.value, self.covered = 0, 0, 0, 0
